### DISCUSSION TITLE
Atom-FG-Salt Replacement Logic

### DISCUSSION THREAD

**Discussion by yulei-chen** [epam/ketcher#2315] [Show and tell]
Hi there👋, this discussion is for tracking all the issues about the replacement logic of atoms, functional groups, and salts & solvents.
These tables were made by other contributors, I just copy/pasted them (got permission).
Feel free to leave a comment or edit this table, if you think there's something wrong or outdated. 

"-" - Single bond
"+" - no bond
t - terminal  
\* - same connection point

## Action: click

| Mouse/canvas | Atom    | FG      | Salt    | Atom-Atom      | (t)FG-*   | Issues |
| ------------ | ------- | ------- | ------- | -------------- | ------------ | ----- |
| **Atom**     | Replace | Replace | Replace | Replace        | Replace      | #2266 |
| **FG**       | Replace | Replace | Replace | Replace        | Replace      | #2326 |
| **Salt**     | Replace | Replace | Replace | Salt+Atom-Atom | Salt+FG-Atom | #2267 #1953 |

## Action: click and drag

| Mouse/canvas | Atom      | FG      | Salt      | Atom-Atom      | (t)FG-*   | Issues  |
| :----------- | --------- | ------- | --------- | -------------- | ------------ | ------ |
| **Atom**     | Atom-Atom | Atom-FG | Atom+Salt | Atom-Atom-Atom | Atom-FG-* | #2276  |
| **FG**       | FG-Atom   | FG-FG   | FG+Salt   | FG-Atom-Atom   | FG-FG-*   | #2278  |
| **Salt**     | Salt+Atom | Salt+FG | Salt+Salt | Salt+Atom-Atom | Salt+FG-* | #2280  |

## Action: move selected

| Mouse/canvas  | Atom    | FG      | Salt    | Atom-Atom | (t)FG-*    |   Issues |
| :------------ | ------- | ------- | ------- | --------- | ------------- | ------- |
| **Atom**      | Replace | Replace | Replace | Replace   | Replace       | #2273   |
| **FG**        | Replace | Replace | Replace | Replace   | Replace       | #2274   |
| **Salt**      | Replace | Replace | Replace | Replace   | Salt +FG-Atom | #2275   |
| **Atom-Atom** | Replace | Replace | Replace | Replace   | Replace       | #2277   |
| **FG-Atom**   | Replace | Replace | Replace | Replace   | Replace       | #2279   |

**Comment by yulei-chen** [comment]
When I was working on #2273, I took some time to figure out the code logic of moving the selected element to other elements.
Here's what I summarized. It would be great if it helps.

![ketcher-merge-2023-03-10-1149](https://user-images.githubusercontent.com/27288153/224218873-da761c0f-5e4a-421f-8785-5bb657907de2.png)

(https://excalidraw.com/#json=lRpjSCXjrdl6MEwDCdXcA,yvanBgfce24lUCaCOLHH9g)

**Comment by kaluginserg** [comment]
An additional investigation of issues with v2.11.0, with moving structures over the canvas and merging structures:

<h3>Legend<h3>
✔ implemented / works
-  not implemented/doesn't work
🚩 bug
❓not clear

<h3>Structure tool</h3>

Tool | Attachment detection for atom | Attachment detection for bound | is SUP SGroup | Group.isCollapsed | Click on the Atom | Click on the Bond | Click and drag from the Atom | Click and drag from the Bond
-- | -- | -- | -- | -- | -- | -- | -- | --
Template “Functional Groups” | ✔ | - | ✔ | ✔ | Merge atom | - | New Bond | -
Template “Salts and Solvents” | ✔ | - | ✔ | ✔ | - | - | Same result | -
Template “Template Library” | ✔ | ✔ | - | - | Merge atom | Merge atom | New Bond | Merged bond
Template “Structure” | ✔ | ✔ | - | - | Merge atom | Merged bond | New Bond | Merged bond
R-Group label | ✔ | - | - | - | Merge atom | - | Same result but with new position | -
Chain | ✔ | ✔ | - | - | -❓ | Change bond type | New Chain | Ignore
Bond | ✔ | ✔ | - | - | New Bond | Change bond type | New Bond | Ignore
Atom & Any atom | ✔ | - | - | - | Merge atom | - | New Bond | -

<h3>Paste tool</h3>

Tool | Attachment detection for atom | Attachment detection for bound | is SUP SGroup | Group.isCollapsed | Click on the Atom | Click on the Bond | Click and drag from the Atom | Click and drag from the Bond
-- | -- | -- | -- | -- | -- | -- | -- | --
Template “Functional Groups” | ✔ | - | ✔ | ✔ | -🚩 | - | Same result but with new position 🚩 | -
Template “Salts and Solvents” | ✔ | - | ✔ | ✔ | - | - | Same result but with new position 🚩 | -
Template “Template Library” | ✔ | ✔ | - | - | Multiple Merge❓ | Multiple Merge❓ | - 🚩 | - 🚩
Template “Structure” | ✔ | ✔ | - | - | Multiple Merge❓ | Multiple Merge❓ | - 🚩 | - 🚩
R-Group label | ✔ | - | - | - | Merge atom | - | Same result but with new position | -
Chain | ✔ | ✔ | - | - | Multiple Merge | Multiple Merge | - 🚩 | - 🚩
Bond | ✔ | ✔ | - | - | Multiple Merge | Multiple Merge | /- | /-
Atom & Any atom | ✔ | - | - | - | Merge atom | - | - 🚩 | -

<h3>Select tool</h3>

Tool | Attachment detection for atom | Attachment detection for bound | is SUP SGroup | SGroup.isCollapsed | Click on the Atom | Click on the Bond | Click and drag from the Atom | Click and drag from the Bond
-- | -- | -- | -- | -- | -- | -- | -- | --
Template “Functional Groups” | ✔ | - | ✔ | -🚩 | Merge atom | - | NOT APPLICABLE | NOT APPLICABLE
Template “Salts and Solvents” | ✔ | - | ✔ | ✔ | - | - | NOT APPLICABLE | NOT APPLICABLE
Template “Template Library” | ✔ | ✔ | - | - | Multiple Merge | Multiple Merge | NOT APPLICABLE | NOT APPLICABLE
Template “Structure” | ✔ | ✔ | - | - | Multiple Merge | Multiple Merge | NOT APPLICABLE | NOT APPLICABLE
R-Group label | ✔ | - | - | - | Merge atom | - | NOT APPLICABLE | NOT APPLICABLE
Chain | ✔ | ✔ | - | - | Multiple Merge | Multiple Merge | NOT APPLICABLE | NOT APPLICABLE
Bond | ✔ | ✔ | - | - | Multiple Merge | Multiple Merge | NOT APPLICABLE | NOT APPLICABLE
Atom & Any atom | ✔ | - | - | - | Merge atom | - | NOT APPLICABLE | NOT APPLICABLE

**Reply by yulei-chen** [reply]
such a great investigation
